guides/hack/02-source-code-fundamentals/37-namespaces.md

```diff
@@ -1,30 +1,39 @@
-A problem encountered when managing large projects is that of avoiding the use of the same name in the same scope for different
-purposes. This is especially problematic in a language that supports modular design and component libraries.  This problem is
-addressed by namespaces.
+A ***namespace*** is a container for a set of (typically related) definitions of classes, interfaces, traits, functions, and constants.
 
-A *namespace* is a container for a set of (typically related) definitions of classes, interfaces, traits, functions, and constants.
+When the same namespace is defined in multiple scripts, and those scripts are combined into the same program, the namespace
+is the union of all individual components.
 
-A namespace may have *sub-namespaces*, where a sub-namespace name shares a common prefix with another namespace. For example, the
-namespace `Graphics` might have sub-namespaces `Graphics\TwoD` and `Graphics\ThreeD`, for two- and three-dimensional facilities,
-respectively. Apart from their common prefix, a namespace and its sub-namespaces have no special relationship. The namespace
-whose prefix is part of a sub-namespace need not actually exist for the sub-namespace to exist. That is, `NS1\Sub` can exist
-without `NS1`.
+The predefined constant `__NAMESPACE__` contains the name of the current namespace.
```

Comment:
Probably should mention:
- `__NAMESPACE__` is primarily for debugging
- `namespace\Foo` (literally lowercase 'namespace') always refers to `Foo` in the current namespace. This can be useful for:
  - clarity, especially e.g. `namespace\map()` is arguably clearer than `map()` in `HH\Lib\Vec` (though it's probabbly better to do a `use namespace HH\Lib\Vec;` and `Vec\map()`, even in that namespace)
  - disambiguating when there is an autoimport

Reply:
Thank you, you taught me something today. I didn't know that use clauses could refer to the current namespace.

Reply:
Is this technically correct?

> For clarity, you can use the word "namespace" and form `namespace\Foo` to refer to properties or methods in the current namespace. For example, an alternative way of referring to `Vec\map()`in `HH\Lib\Vec` is `namespace\map()`. This syntax can be useful when you're importing namespaces with many same-named functions and want to refer to the specific function or method in the scope of the current namespace.

Reply:
Namespaces do not have properties or methods. They have functions, type aliases, classishes, and constants. The umbrella term for those things is entity.

The use for `namespace\map()` is as follows. You are in a file that has a namespace header `namespace HH\Lib\Vec`. The usual way to refer to a function from the same namespace would be by calling `map()`. However, many namespaces declare a map() function (Keyset, Dict). You might want to draw attention and say "I mean map from my current namespace". These two (`map()`, `namespace\map()`) are identical if no use clauses are in play. If you wrote `use function HH\Lib\Dict\map;` in the file, now `map()` resolves to the Dict version. You now need to use the `namespace\map()` variant to say "I don't want you to resolve using that use clause, I actually mean `map()` as declared in the `HH\Lib\Vec` namespace."

Reply:
> Namespaces do not have properties or methods. They have functions, type aliases, classishes, and constants. The umbrella term for those things is entity.
> 
> The use for `namespace\map()` is as follows. You are in a file that has a namespace header `namespace HH\Lib\Vec`. The usual way to refer to a function from the same namespace would be by calling `map()`. However, many namespaces declare a map() function (Keyset, Dict). You might want to draw attention and say "I mean map from my current namespace". These two (`map()`, `namespace\map()`) are identical if no use clauses are in play. If you wrote `use function HH\Lib\Dict\map;` in the file, now `map()` resolves to the Dict version. You now need to use the `namespace\map()` variant to say "I don't want you to resolve using that use clause, I actually mean `map()` as declared in the `HH\Lib\Vec` namespace."

Yes, I should have said constants and functions. Mixing up Class syntax.

Reply:
@lexidor aside from the issue of redeclaring an existing library function, which will raise an error, it looks like there's an error for trying to do the scenario you described above. That is, if my example is correct.
<img width="785" alt="image" src="https://user-images.githubusercontent.com/5179225/162643638-541095bf-16a9-4aff-af63-dc1368a6d736.png">

---

Intent is to add a code example to the docs:
```
namespace HH\Lib\Vec;
use function HH\Lib\Dict\map; // map() resolves to `HH\Lib\Dict\map`

// example implementation of HH\Lib\Vec\map
function map<Tv1, Tv2>(
  Traversable<Tv1> $traversable,
  (function(Tv1)[_]: Tv2) $value_func,
)[ctx $value_func]: vec<Tv2> {
  $result = vec[];
  foreach ($traversable as $value) {
    $result[] = $value_func($value);
  }
  return $result;
}

function multiply_vec_by_two(vec<int> $v): vec<int> {
  // resolves to `HH\Lib\Vec\map`, not `HH\Lib\Dict\map`
  return namespace\map($v, (int $value): int ==> { return 2 * $value; },);
}

<<__EntryPoint>>
function main(): void {
  $simple_vec = vec[1, 2, 3];
  $new_vec = multiply_vec_by_two($simple_vec); // a vec of [2, 4, 6]
}
```

Reply:
You can remove the declaration of Vec\map. If you remove that declaration, and hover over the call namespace\map, you'll see you are calling HH\Lib\Vec\map instead of HH\Lib\Dict\map from the use clause.